test/Dialect/Tpp/tpp-bufferization.mlir

```diff
@@ -779,3 +779,81 @@ func.func @non_conflicting_adds(%arg0: tensor<3x3xf32>, %arg1: tensor<3x3xf32>)
+// CHECK: tpp.gemm ins(%[[ALLOC_1]] : memref<10x10xf32>, %{{.+}} : memref<10x10xf32>, %[[ALLOC]] : memref<10x10xf32>) outs(%[[ALLOC]] : memref<10x10xf32>)
+
+// -----
+
+// A and B are not constants. There is a conflict on %1.
+func.func @gemm_back_to_back(%arg0: tensor<10x10xf32>, %arg1: tensor<10x10xf32>,
+                             %arg2: tensor<10x10xf32>) -> tensor<10x10xf32> {
+  %0 = tensor.empty() : tensor<10x10xf32>
+  %1 = tpp.zero (%0 : tensor<10x10xf32>) -> (tensor<10x10xf32>)
+  %2 = tpp.gemm (%arg0 : tensor<10x10xf32>, %arg1 : tensor<10x10xf32>, %1 : tensor<10x10xf32>) -> (tensor<10x10xf32>)
+  %3 = tpp.gemm (%2 : tensor<10x10xf32>, %arg2 : tensor<10x10xf32>, %1 : tensor<10x10xf32>) -> (tensor<10x10xf32>)
+  return %3 : tensor<10x10xf32>
+}
+
+// CHECK-LABEL: gemm_back_to_back
+// CHECK-SAME: %[[ARG0:.+]]: memref<10x10xf32>, %[[ARG1:.+]]: memref<10x10xf32>, %[[ARG2:.+]]: memref<10x10xf32>
+// CHECK: %[[ALLOC:.+]] = memref.alloc() {{.+}} : memref<10x10xf32>
+// CHECK: tpp.zero ins(%[[ALLOC]] : memref<10x10xf32>) outs(%[[ALLOC]] : memref<10x10xf32>)
+// CHECK: %[[ALLOC_1:.+]] = memref.alloc() {{.+}} : memref<10x10xf32>
+// CHECK: memref.copy %[[ALLOC]], %[[ALLOC_1]] : memref<10x10xf32> to memref<10x10xf32>
+// CHECK: tpp.gemm ins(%[[ARG0]] : memref<10x10xf32>, %[[ARG1]] : memref<10x10xf32>, %[[ALLOC_1]] : memref<10x10xf32>) outs(%[[ALLOC_1]] : memref<10x10xf32>)
+// CHECK: tpp.gemm ins(%[[ALLOC_1]] : memref<10x10xf32>, %[[ARG2]] : memref<10x10xf32>, %[[ALLOC]] : memref<10x10xf32>) outs(%[[ALLOC]] : memref<10x10xf32>)
+
+// -----
+
+// A and B are not constants and C is non-zero. There is a conflict on `arg3`.
+func.func @gemm_back_to_back(%arg0: tensor<10x10xf32>, %arg1: tensor<10x10xf32>,
+                             %arg2: tensor<10x10xf32>, %arg3: tensor<10x10xf32>) -> tensor<10x10xf32> {
+  %2 = tpp.gemm (%arg0 : tensor<10x10xf32>, %arg1 : tensor<10x10xf32>, %arg3 : tensor<10x10xf32>) -> (tensor<10x10xf32>)
+  %3 = tpp.gemm (%2 : tensor<10x10xf32>, %arg2 : tensor<10x10xf32>, %arg3 : tensor<10x10xf32>) -> (tensor<10x10xf32>)
+  return %3 : tensor<10x10xf32>
+}
+
+// CHECK-LABEL: gemm_back_to_back
+// CHECK-SAME:  %[[ARG0:.+]]: memref<10x10xf32>, %[[ARG1:.+]]: memref<10x10xf32>, %[[ARG2:.+]]: memref<10x10xf32>, %[[ARG3:.+]]: memref<10x10xf32>
+// CHECK: %[[ALLOC:.+]] = memref.alloc() {{.+}} : memref<10x10xf32>
+// CHECK: memref.copy %[[ARG3]], %[[ALLOC]] : memref<10x10xf32> to memref<10x10xf32>
+// CHECK: tpp.gemm ins(%[[ARG0]] : memref<10x10xf32>, %[[ARG1]] : memref<10x10xf32>, %[[ALLOC]] : memref<10x10xf32>) outs(%[[ALLOC]] : memref<10x10xf32>)
+// CHECK: tpp.gemm ins(%[[ALLOC]] : memref<10x10xf32>, %[[ARG2]] : memref<10x10xf32>, %[[ARG3]] : memref<10x10xf32>) outs(%[[ARG3]] : memref<10x10xf32>)
+
+// -----
+
+// A and B are not constants and C is non-zero. There are not conflicts.
+func.func @gemm_back_to_back(%arg0: tensor<10x10xf32>, %arg1: tensor<10x10xf32>,
```

Comment:
If you have in mind something like:
```
func.func @gemms(%arg0: tensor<10x10xf32>, %arg1: tensor<10x10xf32>,
                 %arg2: tensor<10x10xf32>, %arg3: tensor<10x10xf32>) -> (tensor<10x10xf32>, tensor<10x10xf32>) {
  %0 = tensor.empty() : tensor<10x10xf32>
  %1 = tpp.zero (%0: tensor<10x10xf32>) -> tensor<10x10xf32>
  %2 = tpp.gemm (%arg0: tensor<10x10xf32>, %arg1: tensor<10x10xf32>, %1: tensor<10x10xf32>) -> tensor<10x10xf32>
  %3 = tpp.gemm (%arg2: tensor<10x10xf32>, %arg3: tensor<10x10xf32>, %1: tensor<10x10xf32>) -> tensor<10x10xf32>
  return %2, %3 : tensor<10x10xf32>, tensor<10x10xf32>
}
```
This will have a conflict too, as %1 has read and write access.

Reply:
Hm, you're right. Let's just add that test with the two allocs and call it a day.